aptos-move/e2e-move-tests/src/tests/governance_updates.rs

```diff
@@ -69,6 +72,77 @@ fn large_transactions() {
     assert!(status.is_discarded());
 }
 
+#[test]
+fn alt_execution_limit_for_gov_proposals() {
+    // This test validates that approved governance scripts automatically gets the
+    // alternate (usually increased) execution limit.
+    let max_gas_regular = 10;
+    let max_gas_gov = 100;
+
+    // Set up the testing environment
+    let mut h = MoveHarness::new();
+
+    let alice = h.new_account_at(AccountAddress::from_hex_literal("0xa11ce").unwrap());
+    let root = h.aptos_framework_account();
+
+    h.modify_gas_schedule(|gas_params| {
+        let txn = &mut gas_params.vm.txn;
+
+        txn.max_execution_gas = Gas::new(10).to_unit_with_params(txn);
+        txn.max_execution_gas_gov = Gas::new(100).to_unit_with_params(txn);
+    });
+    h.set_resource(
+        *root.address(),
+        ApprovedExecutionHashes::struct_tag(),
+        &ApprovedExecutionHashes { entries: vec![] },
+    );
+
+    // Compile the test script, which contains nothing but an infinite loop.
+    let package = BuiltPackage::build(
+        common::test_dir_path("infinite_loop.data/empty_loop_script"),
+        BuildOptions::default(),
+    )
+    .expect("should be able to build package");
+    let script = package
+        .extract_script_code()
+        .pop()
+        .expect("should be able to get script");
+
+    // Execute the script. The amount of gas used should fall within the regular limit.
+    let txn = h.create_script(&alice, script.clone(), vec![], vec![]);
+    let output = h.run_raw(txn);
+    assert_eq!(
+        output.status(),
+        &TransactionStatus::Keep(ExecutionStatus::MiscellaneousError(Some(
+            StatusCode::EXECUTION_LIMIT_REACHED
+        ))),
+    );
+    let gas_used = output.gas_used();
+    assert!(max_gas_regular - 1 <= gas_used && gas_used <= max_gas_regular + 1);
+
+    // Add the hash of the script to the list of approved hashes.
+    h.set_resource(
+        *root.address(),
+        ApprovedExecutionHashes::struct_tag(),
+        &ApprovedExecutionHashes {
+            entries: vec![(0, HashValue::sha3_256_of(&script).to_vec())],
+        },
+    );
+
+    // Execute the script again. This time the amount of gas consumed should be much higher, but
+    // still fall within the alt limit for gov scripts.
+    let txn = h.create_script(&alice, script.clone(), vec![], vec![]);
+    let output = h.run_raw(txn);
+    assert_eq!(
+        output.status(),
+        &TransactionStatus::Keep(ExecutionStatus::MiscellaneousError(Some(
+            StatusCode::EXECUTION_LIMIT_REACHED
+        ))),
+    );
+    let gas_used = output.gas_used();
+    assert!(max_gas_gov - 1 <= gas_used && gas_used <= max_gas_gov + 1);
+}
+
```

Comment:
This is actually a really good point. Although I'm just realizing that it's surprisingly hard to come up with a transaction that costs x amount of gas, without it being extremely fragile (in case we change gas/limits).

For now I think the two infinite loop transactions I have should be good enough -- the second one verifies we are indeed getting the higher execution limit, and it's probably safe to claim that an otherwise successful transaction will also enjoy the same benefits. I will leave a comment saying that we should add the case you mentioned though, if we ever figure out how to do it in a robust way.